### ISSUE TITLE
Should illegal execution contextes be allowed?

### ISSUE THREAD

**Issue by teppolainio** [jordimontana82/fake-xrm-easy#342]
On[ SDK documentation](https://docs.microsoft.com/en-us/dynamics365/customer-engagement/developer/understand-data-context-passed-plugin#input-and-output-parameters) it is said that for example with Delete message Target of InputParameters is EntityReference. I've seen on few occasions developers forgetting this when implementing unit tests and they've used Entity type instead of EntityReference.

Should this kind of errors generally be caught by fake-xrm-easy and raised as exceptions when test is run? 

This particular case for example is pretty clear misunderstanding by developer about how platform execution pipeline works. Most likely one has just forgotten but nevertheless this makes developer to write code which doesn't work on actual environment and this is pretty easy to catch.

**Comment by jordimontana82** [comment]
Hi @teppolainio 

That's great feedback, thanks Teppo.

Yes, I think it would be beneficial to raise exceptions like that, specially for developers who are new to the platform.

Will you fancy a PR?

**Comment by jordimontana82** [comment]
We could possibly add a new method to check illegal execution contexts [at the end of this method](https://github.com/jordimontana82/fake-xrm-easy/blob/master/FakeXrmEasy.Shared/XrmFakedContext.Plugins.cs#L43-L82)

**Comment by teppolainio** [comment]
I'm somewhat busy now but I think I can have time to start with a PR within a month.

What I've been thinking a bit is to have execution context validation done by message executors because valid or invalid execution context pretty much depends from message being executed. What do you think?

**Comment by jordimontana82** [comment]
@teppolainio Yes, I think adding a new method like "bool IsValid()" on the IFakeMessageExecutor interface would be nice, and we could default it to true for existing messages so it doesn't cause regressions, and slowly include these validations.

**Comment by jordimontana82** [comment]
Something like 

    public IsValid(OrganizationRequest request) {

    }

And we update the main entry point in the .Execute mock in XrmFakedContext class to call .IsValid after .CanExecute and before .Execute
